File: rx-codelab/src/main/java/com/jraska/rx/codelab/http/GitHubConverter.java
package com.jraska.rx.codelab.http;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Many combination here to avoid pain with finding proper order of observables.
 */
public class GitHubConverter {
  public static User convert(GitHubUser gitHubUser) {
    boolean isAdmin = gitHubUser.siteAdmin == null ? false : gitHubUser.siteAdmin;
    return new User(gitHubUser.login, gitHubUser.avatarUrl, isAdmin, gitHubUser.htmlUrl);
  }

  public static Repo convert(GitHubRepo gitHubRepo) {
    return new Repo(gitHubRepo.name, gitHubRepo.description,
      gitHubRepo.stargazersCount, gitHubRepo.forks, gitHubRepo.size);
  }

  public static List<Repo> convert(List<GitHubRepo> gitHubRepos) {
    return gitHubRepos.stream()
      .map(GitHubConverter::convert)
      .collect(Collectors.toList());
  }

  public static UserWithRepos convert(GitHubUser gitHubUser, List<GitHubRepo> gitHubRepos) {
    return new UserWithRepos(convert(gitHubUser), convert(gitHubRepos));
  }

  public static UserWithRepos convert(List<Repo> gitHubRepos, User user) {
    return new UserWithRepos(user, gitHubRepos);
  }

  public static UserWithRepos convert(User user, List<Repo> repos) {
    return new UserWithRepos(user, repos);
  }
}

File: rx-codelab/src/main/java/com/jraska/rx/codelab/http/GitHubConverter.kt
package com.jraska.rx.codelab.http

/**
 * Many combination here to avoid pain with finding proper order of observables.
 */
object GitHubConverter {
  fun convert(gitHubUser: GitHubUser): User {
    val isAdmin = gitHubUser.siteAdmin ?: false
    return User(gitHubUser.login, gitHubUser.avatarUrl, isAdmin, gitHubUser.htmlUrl)
  }

  fun convert(gitHubRepo: GitHubRepo): Repo {
    return Repo(gitHubRepo.name, gitHubRepo.description ?: "",
      gitHubRepo.stargazersCount, gitHubRepo.forks, gitHubRepo.size)
  }

  fun convert(gitHubRepos: List<GitHubRepo>): List<Repo> {
    return gitHubRepos.map { convert(it) }
  }

  fun convert(gitHubUser: GitHubUser, gitHubRepos: List<GitHubRepo>): UserWithRepos {
    return UserWithRepos(convert(gitHubUser), convert(gitHubRepos))
  }

  fun convert(gitHubRepos: List<Repo>, user: User): UserWithRepos {
    return UserWithRepos(user, gitHubRepos)
  }

  fun convert(user: User, repos: List<Repo>): UserWithRepos {
    return UserWithRepos(user, repos)
  }
}
